### ISSUE TITLE
Using Sparrow with Tor daemon

### ISSUE THREAD

**Issue by kn0wmad** [sparrowwallet/sparrow#83]
Love the built-in tor support, but this does not play nicely when running the tor daemon for other applications as Sparrow wants port 9050, which is occupied.  Would it be possible to offer to use the existing connection instead of Sparrow's built-in option?

**Comment by craigraw** [comment]
Yes - just configure the proxy settings: proxy on, host localhost, port 9050. Sparrow will prefer this if it is configured.

**Comment by MattDHill** [comment]
There are no proxy settings inside the Core RPC tab. Will inputting them inside the Public Server tab or Private Electrum tab do the trick? Perhaps proxy setting could be abstracted from the various backend options?

**Comment by kn0wmad** [comment]
I ought to have mentioned I'm pointing this at a Bitcoin Core node that is not local to the Sparrow machine.

**Comment by craigraw** [comment]
Unfortunately the bwt project which Sparrow uses internally to connect to Bitcoin Core is waiting on proxy support on one of it's dependencies, the rust-bitcoincore-rpc project. Essentially, that project needs to update to a newer version of it's jsonrpc library for Tor proxy support to be possible. You can see the open PR for this here: https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/154

This is a good place to track this, so cc'ing @shesek (who is aware of the above)

**Comment by kn0wmad** [comment]
May I ask why Sparrow is using BWT to connect to Bitcoin Core?  Would it be possible to connect directly via RPC as Specter and FullyNoded do, for example?

**Comment by craigraw** [comment]
Sparrow uses the Electrum server protocol to remain independent from the Bitcoin Core wallet. Using the Bitcoin Core RPC directly would effectively mean recreating BWT in some way in order to maintain this independence. If Bitcoin Core ever decided to build an address-based index this would not be an issue - unfortunately this seems very unlikely.

**Comment by ProofOfKeags** [comment]
OK so as I understand it, this functionality is delegated to BWT itself, which is packaged as a binary inside of Sparrow? I think I can get the PR above or equivalent functionality into the sparrow stack, I just need to know what the dependency chain is. Swappable transports were implemented in rust's jsonrpc library since 0.12.0. I was just able to create a PR [here]( https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/180) to get the new transport machinery into the actual rpc library. But it sounds like from up thread that there remains the issue of after this gets merged we need to update BWT to use an optional tor transport. Correct?

**Comment by craigraw** [comment]
> OK so as I understand it, this functionality is delegated to BWT itself, which is packaged as a binary inside of Sparrow?

Yes correct. Updating BWT shouldn't be an issue, but @shesek would be able to comment better on this.

**Comment by 6102bitcoin** [comment]
Outstanding Action: Discuss with @shesek the possibility of an "optional tor transport" in bwt
Proposed Priority: Low

**Comment by ProofOfKeags** [comment]
I would prefer that it wasn't low priority and am willing to give dev resources to actually writing the code to do so. However, the code is going to be contingent upon getting [this](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/180) merged first.

**Comment by craigraw** [comment]
I regard this as Medium Priority (although I'm not recommending this approach - I don't think opening Bitcoin Core RPC over Tor is a great idea for privacy reasons). However I certainly see the need for this, and I know @shesek is keen to implement it in BWT once the necessary support has been added to rust-bitcoincorerpc.

**Comment by ProofOfKeags** [comment]
What privacy leak are you concerned about with RPC over Tor?

**Comment by craigraw** [comment]
The issue is that most places the .onion address and RPC auth will be stored are not password protected - config files etc. Learning the RPC .onion address means you can easily extract all the watched addresses and their balances. Worse, you can do this remotely, which makes it even easier than scanning the Bitcoin Core wallet files themselves.

That said, some many not consider this an attack vector they are concerned about - it's one wouldn't be happy with though.

**Comment by ProofOfKeags** [comment]
I am not sure where you get the idea that RPC over Tor is not password
protected but somehow RPC over not-Tor is better. In every case I’ve ever
seen, Core has RPC credentials set and even if those aren’t encrypted on
disk, I’m not sure how learning the onion address grants access to the RPC
credentials.

Am I misunderstanding your claim?

**Comment by craigraw** [comment]
RPC over not-Tor is generally limited to local network attacks, which is not true for RPC over Tor. RPC credentials are usually not encrypted on disk in config files etc.

The standing advice has been to not expose RPC beyond localhost: https://bitcoin.stackexchange.com/questions/69080/rpc-json-outside-localhost

**Comment by ProofOfKeags** [comment]
Ok this isn’t anything to do with Tor but is an indictment of any kind of
remote node. Unfortunately this seems unavoidable if you want to have
multiple wallets on multiple devices and don’t want to independently sync
them. It seems that the way to deal with what you’re talking about is to
not allow external nodes of any kind. Personally I think limiting sparrow
to only be usable with an internal node is not a great idea. People want to
be able to use their existing nodes for their wallets. I really like
sparrow and the only barrier to using it is being able to input a remote
node.

**Comment by craigraw** [comment]
> Ok this isn’t anything to do with Tor but is an indictment of any kind of remote node.

Electrum servers don't carry this risk, since they index the entire blockchain, and you can't learn anything about a particular wallet by connecting to them.

> I really like sparrow and the only barrier to using it is being able to input a remote node.

Thanks - and as I said before I'm keen to support this functionality. Just wanted to point out the risks. Let's make it happen.

**Comment by ProofOfKeags** [comment]
Understood. Yes, contingent on getting a valid Authenticated RPC connection
to Core you can absolutely have a privacy leak. I’m primarily contesting
the “ease” of that attack in the presence of Torv3 (blinded keys keep onion
addrs undiscoverable). The link is also encrypted so network actors have no
insight. Finally you’d have to get read access to the disk of the remote
node to get the credentials to Auth to the RPC.

I misunderstood your original statement. Glad that you’re willing to merge
something that would implement this. I’ll steward the requisite PRs in the
dependencies and revisit here when the deps are ready.

😎

**Comment by nyxnor** [comment]
I express my interest on this feature, but it works with electrum already, so if possible to mention RPC on the title.
I agree with https://github.com/sparrowwallet/sparrow/issues/83#issuecomment-846043800

**Comment by craigraw** [comment]
@ProofOfKeags any progress on this front?

**Comment by kn0wmad** [comment]
> @ProofOfKeags any progress on this front?

I believe this is stuck in the dependencies still.  We are approaching a release that allows archival nodes and will advise our users to go that route to use Sparrow when complete.  Until the dependencies get caught up pruned nodes will be unable to use Sparrow unfortunately

**Comment by craigraw** [comment]
> Until the dependencies get caught up pruned nodes will be unable to use Sparrow unfortunately

Thanks, but I'm a little confused - is use of RPC over Tor in rust-bitcoincorerpc connected to node pruning in some way?

**Comment by ProofOfKeags** [comment]
https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/180 has been merged.

The library was released at 0.14.0 5 days ago which includes the changes for swappable transports. So this should allow for sending over a sock5 tor connection to the core instance in question.

> is use of RPC over Tor in rust-bitcoincorerpc connected to node pruning in some way?

Not at all. They are separate concerns. Though both of them affect *our* current users. But this issue is strictly about remote node access over tor.

**Comment by shesek** [comment]
> [rust-bitcoin/rust-bitcoincore-rpc#180](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/180) has been merged.

Sweet! :tada: Will be looking into adding SOCKS proxy support to BWT, probably in the coming days.

**Comment by 6102bitcoin** [comment]
Should this issue be closed?

**Comment by craigraw** [comment]
Indeed it should. Thanks for the upstream work.

**Comment by kn0wmad** [comment]
> > [rust-bitcoin/rust-bitcoincore-rpc#180](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/180) has been merged.
> 
> Sweet! tada Will be looking into adding SOCKS proxy support to BWT, probably in the coming days.

Sorry to necrobump, was this support added to BWT?

**Comment by shesek** [comment]
> Sorry to necrobump, was this support added to BWT?

Yes, SOCKS5 proxy support was added to bwt (https://github.com/bwt-dev/bwt/issues/21, https://github.com/bwt-dev/bwt/commit/97f9edfc29932e4d3f33751fa967d0af9d60bccc)

And to Sparrow in v1.51: https://github.com/sparrowwallet/sparrow/releases/tag/1.5.1

**Comment by kn0wmad** [comment]
Cheers, take care
